core/domain/exp_domain.py

```diff
@@ -3060,90 +3169,1468 @@ def _convert_states_v51_dict_to_v52_dict(
+        """Helper function for NumericInput and FractionInput interaction,
+        Checks whether the ranges of rules enclosed or not
+
+        Args:
+            range_compare_for: RangeVariableDict. It represents the variable for
+                which we have to check the range. It keeps track of each rule's
+                ans group index, rule spec index, lower bound, upper bound,
+                lb inclusive, ub inclusive.
+            range_compare_with: RangeVariableDict. It is the variable to which
+                the range is compared. It keeps track of other rule's
+                ans group index, rule spec index, lower bound, upper bound,
+                lb inclusive, ub inclusive.
+
+        Returns:
+            is_enclosed: bool. Returns True if both rule's ranges are enclosed.
+        """
+        if (
+            range_compare_with['lower_bound'] is None or
+            range_compare_for['lower_bound'] is None or
+            range_compare_with['upper_bound'] is None or
+            range_compare_for['upper_bound'] is None
+        ):
+            return False
+        lb_satisfied = (
+            range_compare_with['lower_bound'] < range_compare_for[
+                'lower_bound'] or
+            (
+                range_compare_with['lower_bound'] == range_compare_for[
+                    'lower_bound'] and
+                (
+                    not range_compare_for['lb_inclusive'] or
+                    range_compare_with['lb_inclusive']
+                )
+            )
+        )
+        ub_satisfied = (
+            range_compare_with['upper_bound'] > range_compare_for[
+                'upper_bound'] or
+            (
+                range_compare_with['upper_bound'] == range_compare_for[
+                    'upper_bound'] and
+                (
+                    not range_compare_for['ub_inclusive'] or
+                    range_compare_with['ub_inclusive']
                 )
+            )
+        )
+        is_enclosed = lb_satisfied and ub_satisfied
+        return is_enclosed
+
+    @classmethod
+    def _should_check_range_criteria(
+        cls,
+        earlier_rule: state_domain.RuleSpecDict,
+        later_rule: state_domain.RuleSpecDict
+    ) -> bool:
+        """Helper function for FractionInput interaction,
+        Checks the range criteria between two rules by comparing their
+        rule type
+
+        Args:
+            earlier_rule: state_domain.RuleSpecDict. Previous rule.
+            later_rule: state_domain.RuleSpecDict. Current rule.
+
+        Returns:
+            bool. Returns True if the rules passes the range criteria check.
+        """
+        if earlier_rule['rule_type'] in (
+            'HasDenominatorEqualTo', 'IsEquivalentTo', 'IsLessThan'
+            'IsEquivalentToAndInSimplestForm', 'IsGreaterThan'
+        ):
+            return True
+        return later_rule['rule_type'] in (
+            'HasDenominatorEqualTo', 'IsLessThan', 'IsGreaterThan'
+        )
```

Comment:
I already reviewed these in the other PR, similar changes apply.

Reply:
applies to some functions below too

Reply:
Done